lib/ansible/modules/windows/win_data_deduplication.ps1

```diff
@@ -0,0 +1,121 @@
+#!powershell
+
+# Copyright: 2019, rnsc(@rnsc) <github@rnsc.be>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt
+
+#Requires -Module Ansible.ModuleUtils.Legacy
+#AnsibleRequires -OSVersion 6.3
+
+$params = Parse-Args $args -supports_check_mode $true
+
+$drive_letter = Get-AnsibleParam -obj $params -name 'drive_letter' -type 'str' -failifempty $true
+$state = Get-AnsibleParam -obj $params -name 'state' -type 'str' -default 'present'
+$settings = Get-AnsibleParam -obj $params -name 'settings' -type 'dict' -default $null
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type 'bool' -default $false
+
+$result = @{
+  changed = $false
+  reboot_required = $false
+  msg = ""
+}
+
+function Set-DataDeduplication($volume, $state, $settings, $dedup_job) {
+
+  $current_state = 'absent'
+
+  try {
+    $dedup_info = Get-DedupVolume -Volume "$($volume.DriveLetter):"
+  } catch {
+    $dedup_info = $null
+  }
+
+  if($null -ne $dedup_info) {
+    if ($dedup_info.Enabled) {
+      $current_state = 'present'
+    }
+  } else {
+    $current_state = $null
+  }
+
+  if ( $state -ne $current_state ) {
+    if( -not $check_mode) {
+      if($state -eq 'present') {
+        # Enable-DedupVolume -Volume <String>
+        Enable-DedupVolume -Volume "$($volume.DriveLetter):"
+      } elseif ($state -eq 'absent') {
+        Disable-DedupVolume -Volume "$($volume.DriveLetter):"
+      } else {
+        Fail-Json $result "Unsupported state option, it can only be 'present' or 'absent'."
+      }
+    }
+    $result.msg += " Deduplication on volume: "+$volume.DriveLetter+" set to: $state"
+    $result.changed = $true
+  }
+
+  if ($state -eq 'present') {
+    if ($null -ne $settings) {
+      Set-DataDedupJobSettings -volume $volume -settings $settings
+    }
+  }
+}
+
+function Set-DataDedupJobSettings ($volume, $settings) {
+
+  try {
+    $dedup_info = Get-DedupVolume -Volume "$($volume.DriveLetter):"
+  } catch {
+    $dedup_info = $null
```

Comment:
So you are setting this to `$null` but then have no checks for if it is `$null` later on? One thing you do need to consider is if you are;

* Running in check mode
* Dedup is not enabled (or missing) but `state: present` is set
* You have settings you want to set

In this scenario `Get-DedupVolume` may return `$null` causing the below to fail. You would need to handle this scenario, probably by just checking if we are in check mode and we can get the settings already.

Reply:
Yes, this part is a bit confusing, I'll take a look at what I can do to fix this.

Reply:
The check is actually needed in case data dedup was never setup on the drive